tensorflow/contrib/metrics/python/ops/confusion_matrix_ops.py

```diff
@@ -0,0 +1,99 @@
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from tensorflow.python.framework import ops
+from tensorflow.python.framework import dtypes
+from tensorflow.python.ops import array_ops
+from tensorflow.python.ops import math_ops
+from tensorflow.python.ops import sparse_ops
+
+
+"""Confusion matrix related metrics."""
+
+
+def confusion_matrix(predictions, targets, num_classes=None, name=None):
+  """Computes the confusion matrix from predictions and targets
+
+  Calculate the Confusion Matrix for a pair of prediction and
+  target 1-D int arrays.
+
+  Considering a prediction array such as: `[1, 2, 3]`
+  And a target array such as: `[2, 2, 3]`
+
+  The confusion matrix returned would be the following one:
+      [[0, 0, 0]
+       [0, 1, 0]
+       [0, 1, 0]
+       [0, 0, 1]]
+
+  Where the matrix rows represent the prediction labels and the columns
+  represents the target labels. The confusion matrix is always a 2-D array
+  of shape [n, n], where n is the number of valid labels for a given
+  classification task. Both prediction and target must be 1-D arrays of
+  the same shape in order for this function to work.
+
+  Args:
+    predictions: A 1-D array represeting the predictions for a given
+                 classification.
+    targets: A 1-D represeting the real labels for the classification task.
+    num_classes: The possible number of labels the classification task can
+                 have. If this value is not provided, it will be calculated
+                 using both predictions and targets array.
+
+  Returns:
+    A l X l matrix represeting the confusion matrix, where l in the number of
+    possible labels in the classification task.
+
+  Raises:
+    ValueError: If both predictions and targets are not 1-D vectors and do not
+                have the same size.
+  """
+  predictions_tensor = ops.convert_to_tensor(
+    predictions, name="inputs_tensor", dtype=dtypes.int64)
+  targets_tensor = ops.convert_to_tensor(
+    targets, name="targets_tensor", dtype=dtypes.int64)
+
+  predictions_shape = predictions_tensor.get_shape()
+  targets_shape = targets_tensor.get_shape()
+
+  if predictions_shape.ndims != 1:
+    raise ValueError(
+      "targets is not a 1-D vector")
+
+  if targets_shape.ndims != 1:
+    raise ValueError(
+      "targets is not a 1-D vector")
+
+  if predictions_shape.num_elements != targets_shape.num_elements:
+    raise ValueError(
+      "predictions and targets must be the same size")
+
+  if num_classes is None:
+    num_classes = math_ops.maximum(math_ops.reduce_max(predictions_tensor),
+                                   math_ops.reduce_max(targets_tensor)) + 1
+
+  shape = array_ops.pack([num_classes, num_classes])
+  indices = array_ops.transpose(
+    array_ops.pack([predictions_tensor, targets_tensor]))
+  values = array_ops.ones(shape=[indices.get_shape()[0], ], dtype=dtypes.int32)
```

Comment:
I think this approach has one problem. In the SparseTensor documentation it is stated that the indices array must have a shape [N, ndims] and values must have a shape of [N]. I believe that this N variable is not always equal to the num_classes one.